### PR TITLE
Added PostgreSQL support

### PR DESCRIPTION
**Pull request by andreybolonin**
Some fixes to enable very important PostgreSQL support.

Why is so important? 

My simple example - on VPS, MySQL 5.7 eat all of RAM memory. And VPS crashed :(

Ok.

Because Pagekit connect to database through Doctrine DBAL, 
we should not many problems with adding some 
another DB (PostgreSQL or some another db).

I am not pro in JS. Say me plz, what we should fix to registration new pagekit copy works fine.

### CODE REVIEW

File: app/installer/app/views/installer.js [hunk 1]
```diff
@@ -16,7 +16,15 @@ var Installer = {
...
-        this.$set('config.database', {default: this.sqlite ? 'sqlite' : 'mysql'})
+        if (this.sqlite) {
+            db = 'sqlite';
+        } else if (this.mysql) {
+            db = 'mysql';
+        } else if (this.pgsql) {
+            db = 'pgsql';
+        }
+        this.$set('config.database', {default: db})
+        // this.$set('config.database', {default: this.sqlite ? 'sqlite' : 'mysql'})
```
Comment (jdegger) [inline app/installer/app/views/installer.js]: Maybe remove this commented line from the patch? Makes the source code less tidy.

File: app/installer/views/installer.php [hunk 2]
```diff
@@ -123,6 +124,45 @@
                                     </div>
                                 </div>
                             </div>
+                            <div class="uk-form-row" v-if="config.database.default == 'pgsql'">
+                                <div class="uk-form-row">
+                                    <label for="form-mysql-dbhost" class="uk-form-label">{{ 'Hostname' | trans }}</label>
+                                    <div class="uk-form-controls">
+                                        <input id="form-mysql-dbhost" class="uk-width-1-1" type="text" name="host" value="localhost" v-model="config.database.connections.mysql.host" v-validate:required>
```
Comment (jdegger) [inline app/installer/views/installer.php]: It feels off that the `.mysql` configuration variables are used. We should either rename that to something more general or use separate configuration variables?

File: app/modules/config/src/ConfigManager.php [hunk 3]
```diff
@@ -97,6 +98,8 @@ public function set($name, $config)
 
             if ($this->connection->getDatabasePlatform() instanceof MySqlPlatform) {
                 $this->connection->executeQuery("INSERT INTO {$this->table} (name, value) VALUES (:name, :value) ON DUPLICATE KEY UPDATE value = :value", $data);
+            } elseif ($this->connection->getDatabasePlatform() instanceof PostgreSqlPlatform) {
+                $this->connection->executeQuery("INSERT INTO {$this->table} (name, value) VALUES (:name, :value) ON DUPLICATE KEY UPDATE value = :value", $data);
```
Comment (jdegger) [inline app/modules/config/src/ConfigManager.php]: Maybe merge the first two ifs since they execute the same code

File: app/modules/session/src/Handler/DatabaseSessionHandler.php [hunk 4]
```diff
@@ -142,6 +143,11 @@ protected function getMergeSql()
         if ($platform instanceof MySqlPlatform) {
             return "INSERT INTO {$this->table} (id, data, time) VALUES (:id, :data, :time) "
             . "ON DUPLICATE KEY UPDATE data = VALUES(data), time = CASE WHEN time = :time THEN (VALUES(time) + INTERVAL 1 SECOND) ELSE VALUES(time) END";
+
+        } elseif ($platform instanceof PostgreSqlPlatform) {
+            return "INSERT INTO {$this->table} (id, data, time) VALUES (:id, :data, :time) "
+            . "ON DUPLICATE KEY UPDATE data = VALUES(data), time = CASE WHEN time = :time THEN (VALUES(time) + INTERVAL 1 SECOND) ELSE VALUES(time) END";
+
```
Comment (jdegger) [inline app/modules/session/src/Handler/DatabaseSessionHandler.php]: Should also be merged I think, we wouldn't expect differences between those lines of code right?

**Comment by jdegger** [comment]
Thank you for your efforts @andreybolonin! I really want postgres support :)

@MalteScharenberg what do you think?

**Comment by MalteScharenberg** [comment]
I like the idea of supporting a verity of database platforms. The only problem with this is that all of them need to be maintained and it must be ensured that all platforms behave exactly the same. 

More platforms means more testing and conceivably necessary adaptations (e.g. in the past Pagekits SQLite Driver behaved slightly differently to Pagekits MySql driver). 

Thats why I'm thinking that this needs to be well considered (and tested) before merged into core and currently there is a lot on Pagekits TODO list with higher priority.

@andreybolonin Whats about SQLite as a lightweight alternative to MySQL?

**Comment by jdegger** [comment]
Implement unit tests to make sure the functionality all works as expected. If the unit tests pass for all drivers it should be easier to maintain those. Is this something that's been considered?

**Comment by MalteScharenberg** [comment]
Pagekit has Unit tests but they are not well maintained (an another thing on the TODO list :) ) and especially Unit tests for databases are not the easiest ones since you have to stub the specific behaviour of every database and there are always edge case. 

I'm not saying that this is not possible. I'm just saying that this needs time, because this is a change at Pagekits foundations.

**Comment by jdegger** [comment]
Understandable. I mean, the PR is nice but there are still some improvements to be made in my opinion. Making sure everything works as expected is very important.

Is it an idea that this is embedded in a feature branch for testing? Also is there are a roadmap for the things you guys consider high priority? I don't want to bother you with things which are not relevant at the moment.

**Comment by andreybolonin** [comment]
@MalteScharenberg To be honest, I don`t experience with SQLite.

But ubuntu sever is updated. And mysql 5.6 memory leak have been fixed.

But, I think that PostgreSQL support is anyway very profitable option.

And what about phpunit tests for databases? They exist in pagekit in this time? :)

**Comment by artjomsimon** [comment]
Just wanted to play around with Pagekit, and for me, PostrgreSQL is a prerequisite.

Would love to see Postgres support integrated

@andreybolonin thank you for this patch, do you have time to incorporate @jdegger 's [suggestions](https://github.com/pagekit/pagekit/pull/708/files?diff=unified#r72413121)?

**Comment by lenovouser** [comment]
:+1: I'd love to see this feature being implemented in Pagekit too!

**Comment by nbpalomino** [comment]
+1
